File: DemoServer/Demos/Advanced/SetBasedScripted.cs
using System.Linq;
using DemoServer.Controllers;
using DemoServer.Entities;
using DemoServer.Helpers;
using Microsoft.AspNetCore.Mvc;
using Raven.Client.Documents.Operations;
using Raven.Client.Documents.Queries;

namespace DemoServer.Demos.Advanced
{
    public partial class AdvancedController : BaseController
    {
        [HttpGet]
        [Route("/advanced/setBasedScripted")]
        [Demo("Set Based Scripted", DemoOutputType.Flatten, demoOrder: 240)]
        public object SetBasedScripted(string employee = "employees/1", int discount = 5)
        {
            var updateByIndex = DocumentStoreHolder.Store.Operations.Send(new PatchByQueryOperation(
                new IndexQuery
                {
                    Query = @"
FROM INDEX 'Orders/Totals' 
WHERE Employee = $emp 
UPDATE {
    for(var i = 0; i < this.Lines.length; i++)
    {
        this.Lines[i].Discount = Math.max(this.Lines[i].Discount || 0, args.discount);
    }
}",
                    QueryParameters = new Raven.Client.Parameters()
                    {
                        ["emp"] = employee,
                        ["discount"] = discount
                    }
                }));

            updateByIndex.WaitForCompletion();

            using (var session = DocumentStoreHolder.Store.OpenSession())
            {
                var results = session
                    .Query<Order>()
                    .Customize(x => x.WaitForNonStaleResults())
                    .Where(x => x.Employee == employee)
                    .ToList();

                return (results);
            }
        }
    }
}

File: DemoServer/PythonDemo/demo_methods/Advanced/SetBasedScripted.py
from demo.holder import DocumentStoreHolder
from pyravendb.data.query import IndexQuery
from pyravendb.raven_operations.operations import QueryOperationOptions, PatchByQueryOperation
from demo.entities import Order


class SetBasedScripted(object):
    @staticmethod
    def set_based_scripted(employee="employees/1", discount=5):
        query = ("FROM INDEX 'Orders/Totals' "
                 "WHERE Employee = $emp "
                 "UPDATE { for(var i = 0; i < this.Lines.length; i++) { "
                 "this.Lines[i].Discount = Math.max(this.Lines[i].Discount || 0, args.discount);}}")

        query_parameters = {"emp": employee, "discount": discount}

        patch_operation = PatchByQueryOperation(
            IndexQuery(query=query, query_parameters=query_parameters),
            options=QueryOperationOptions(allow_stale=False))

        store = DocumentStoreHolder.get_store()
        result = store.operations.send(patch_operation)
        store.operations.wait_for_operation_complete(result['operation_id'])

        with DocumentStoreHolder.get_store().open_session() as session:
            results = list(
                session.query(object_type=Order,
                              wait_for_non_stale_results=True).where(Employee=employee))

            return results
